src/device/pointer.jl

```diff
@@ -80,35 +84,25 @@ Base.convert(::Type{DevicePtr{T}}, p::DevicePtr{U,A}) where {T,U,A} = Base.unsaf
 
 ## limited pointer arithmetic & comparison
 
-Base.:(==)(a::DevicePtr, b::DevicePtr) = pointer(a) == pointer(b) && addrspace(a) == addrspace(b)
+isequal(x::DevicePtr, y::DevicePtr) = (x === y) && addrspace(x) == addrspace(y)
+isless(x::DevicePtr{T,A}, y::DevicePtr{T,A}) where {T,A<:AddressSpace} = x < y
 
-Base.isless(x::DevicePtr, y::DevicePtr) = Base.isless(pointer(x), pointer(y))
-Base.:(-)(x::DevicePtr, y::DevicePtr)   = pointer(x) - pointer(y)
+Base.:(==)(x::DevicePtr, y::DevicePtr) = UInt(x) == UInt(y) && addrspace(x) == addrspace(y)
+Base.:(<)(x::DevicePtr,  y::DevicePtr) = UInt(x) < UInt(y)
+Base.:(-)(x::DevicePtr,  y::DevicePtr) = UInt(x) - UInt(y)
 
-Base.:(+)(x::DevicePtr{T,A}, y::Integer) where {T,A} = DevicePtr{T,A}(pointer(x) + y)
-Base.:(-)(x::DevicePtr{T,A}, y::Integer) where {T,A} = DevicePtr{T,A}(pointer(x) - y)
+Base.:(+)(x::DevicePtr, y::Integer) = oftype(x, Base.add_ptr(UInt(x), (y % UInt) % UInt))
+Base.:(-)(x::DevicePtr, y::Integer) = oftype(x, Base.sub_ptr(UInt(x), (y % UInt) % UInt))
 Base.:(+)(x::Integer, y::DevicePtr) = y + x
 
-
 ## memory operations
 
-@static if Base.libllvm_version < v"7.0"
-    # Old values (LLVM 6)
-    Base.convert(::Type{Int}, ::Type{AS.Private})  = 0
-    Base.convert(::Type{Int}, ::Type{AS.Global})   = 1
-    Base.convert(::Type{Int}, ::Type{AS.Constant}) = 2
-    Base.convert(::Type{Int}, ::Type{AS.Local})    = 3
-    Base.convert(::Type{Int}, ::Type{AS.Generic})  = 4
-    Base.convert(::Type{Int}, ::Type{AS.Region})   = 5
-else
-    # New values (LLVM 7+)
-    Base.convert(::Type{Int}, ::Type{AS.Generic})  = 0
-    Base.convert(::Type{Int}, ::Type{AS.Global})   = 1
-    Base.convert(::Type{Int}, ::Type{AS.Region})   = 2
-    Base.convert(::Type{Int}, ::Type{AS.Local})    = 3
-    Base.convert(::Type{Int}, ::Type{AS.Constant}) = 4
-    Base.convert(::Type{Int}, ::Type{AS.Private})  = 5
-end
+Base.convert(::Type{Int}, ::Type{AS.Generic})  = 0
```

Comment:
Same thing here are these the right values for HSA?

Reply:
Good catch!